### PR TITLE
[hostcfgd]: Add Ability To Configure Feature During Run-time

### PR DESCRIPTION
**Pull request by tahmed-dev**
Features may be enabled/disabled for the same topology based on run-time
configuration. This PR adds the ability to enable/disable feature based
on config db data.

singed-off-by: Tamer Ahmed <tamer.ahmed@microsoft.com>

**- Why I did it**
Dual ToR topology have mux docker while non single ToR topology does not and hence the need to disable feature during run-time.

**- How I did it**
Added the ability to pass Jinja2 template to hostcfgd. The template could be as follows:
`"{% if 'subtype' in DEVICE_METADATA['localhost'] and DEVICE_METADATA['localhost']['subtype'] == 'DualToR' %}enabled{% else %}always_disabled{% endif %}"`

**- How to verify it**
On Target:

Single ToR
```
admin@str-s6000-acs-14:~$ show feature statu
Feature     State            AutoRestart     SystemState    UpdateTime           ContainerId    Version                            SetOwner    CurrentOwner    RemoteState
----------  ---------------  --------------  -------------  -------------------  -------------  ---------------------------------  ----------  --------------  -------------
acms        enabled          enabled         up             2021-02-05 21:21:33  acms                                              local       local           none
bgp         enabled          enabled         up             2021-02-05 21:21:36  bgp                                               local       local           none
database    always_enabled   always_enabled                                                                                        local
dhcp_relay  enabled          enabled         up             2021-02-05 21:22:37  dhcp_relay     mux-debug.0-dirty-20210204.123047  kube        local
lldp        enabled          enabled         up             2021-02-05 21:22:37  lldp           mux-debug.0-dirty-20210204.123047  kube        local
mux         always_disabled  enabled         down           2021-02-05 21:26:52                                                    local       none            none
pmon        enabled          enabled         up             2021-02-05 23:08:07  pmon           mux-debug.0-dirty-20210204.123047  kube        local
radv        enabled          enabled         up             2021-02-05 21:22:32  radv           mux-debug.0-dirty-20210204.123047  kube        local
snmp        enabled          enabled         up             2021-02-05 23:06:45  snmp           mux-debug.0-dirty-20210204.123047  kube        local
swss        enabled          enabled         up             2021-02-05 21:21:50  swss                                              local       local           none
syncd       enabled          enabled         up             2021-02-05 21:22:02  syncd                                             local       local           none
teamd       enabled          enabled         up             2021-02-05 21:22:01  teamd                                             local       local           none
telemetry   enabled          enabled         up             2021-02-05 21:24:12  telemetry      mux-debug.0-dirty-20210204.123047  kube        local
```

Dual ToR
```
admin@STR43-0101-0101-01LT0:~$ show feature status 
Feature     State           AutoRestart     SystemState    UpdateTime           ContainerId    Version                            SetOwner    CurrentOwner    RemoteState
----------  --------------  --------------  -------------  -------------------  -------------  ---------------------------------  ----------  --------------  -------------
acms        enabled         enabled         up             2021-02-05 19:22:21  acms                                              local       local           none
bgp         enabled         enabled         up             2021-02-05 20:03:28  bgp                                               local       local           none
database    always_enabled  always_enabled                                                                                        local
dhcp_relay  enabled         enabled         up             2021-02-05 20:03:44  dhcp_relay     mux-debug.0-dirty-20210204.123047  kube        local
lldp        enabled         enabled         up             2021-02-05 20:03:40  lldp           mux-debug.0-dirty-20210204.123047  kube        local
mux         enabled         enabled         up             2021-02-05 20:05:03  mux                                               local       local           none
pmon        enabled         enabled         up             2021-02-05 20:03:30  pmon           mux-debug.0-dirty-20210204.123047  kube        local
radv        enabled         enabled         up             2021-02-05 20:03:36  radv           mux-debug.0-dirty-20210204.123047  kube        local
snmp        enabled         enabled         up             2021-02-05 20:05:09  snmp           mux-debug.0-dirty-20210204.123047  kube        local
swss        enabled         enabled         up             2021-02-05 20:03:26  swss                                              local       local           none
syncd       enabled         enabled         up             2021-02-05 20:03:28  syncd                                             local       local           none
teamd       enabled         enabled         up             2021-02-05 20:03:35  teamd                                             local       local           none
telemetry   enabled         enabled         up             2021-02-05 20:04:24  telemetry      mux-debug.0-dirty-20210204.123047  kube        local
```

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [x] 202012

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by jleveque** [comment]
Can you provide an example/description of how this template is to be passed to hostcfgd?

**Comment by tahmed-dev** [comment]
> Can you provide an example/description of how this template is to be passed to hostcfgd?

in the "How I did it" section above, there is an example template.

**Comment by jleveque** [comment]
> > Can you provide an example/description of how this template is to be passed to hostcfgd?
> 
> in the "How I did it" section above, there is an example template.

I see the template. Just want to understand how it is to be passed to the daemon.

**Comment by tahmed-dev** [comment]
> > > Can you provide an example/description of how this template is to be passed to hostcfgd?
> > 
> > 
> > in the "How I did it" section above, there is an example template.
> 
> I see the template. Just want to understand how it is to be passed to the daemon.

The template gets written to the config_db past of the bootstrapping when init_cfg.json is loaded to the config_db. hostcfgd will consume it from there.

**Comment by jleveque** [comment]
> The template gets written to the config_db past of the bootstrapping when init_cfg.json is loaded to the config_db. hostcfgd will consume it from there.

The template gets written as a bogus feature state?

**Comment by tahmed-dev** [comment]
> > The template gets written to the config_db past of the bootstrapping when init_cfg.json is loaded to the config_db. hostcfgd will consume it from there.
> 
> The template gets written as a bogus feature state?

Initially, this [line](https://github.com/Azure/sonic-buildimage/pull/6700/files#diff-27d10088e7a11c3ae41f16e5272d96926179ae179c61d6fc32a25d0cfa40907eR435) should write the rendered value.

**Comment by jleveque** [comment]
Thanks for the details. While this is a clever solution, I feel it is a bit hacky and complex. Can we not do something simpler, like when hostcfgd first starts, if it detects the subtype is `DualToR`, It simply changes the state to `enabled`?

**Comment by tahmed-dev** [comment]
> Thanks for the details. While this is a clever solution, I feel it is a bit hacky and complex. Can we not do something simpler, like when hostcfgd first starts, if it detects the subtype is `DualToR`, It simply changes the state to `enabled`?

Welcome! I was kind of thinking of more generic approach as if there are more features as such, we would ended filling up hostcfgd with special cases instead of caring about the feature attributes. I open to other solutions.

File: src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            Compares Config DB table (FEATURE) with expected output table
+
+            Args:
+                table(dict): Current Config Db table
+                expected_table(dict): Expected Config Db table
+
+            Returns:
+                None
+        """
+        isEqual = len(table) == len(expected_table)
```
Comment (jleveque) [inline src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py:45]: Variable names in Python code should be `lowercase_with_underscores` (e.g., `is_equal`)

File: src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            Args:
+                name(str): test name
+                test_data(dict): test data which contains initial Config Db tables, and expected results
+
+            Returns:
+                None
+        """
+        MockConfigDb.set_config_db(testData["config_db"])
+        with mock.patch("hostcfgd.subprocess") as mockSubprocess:
+            hostConfigDaemon = hostcfgd.HostConfigDaemon()
```
Comment (jleveque) [inline src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py:72]: Variable names in Python code should be `lowercase_with_underscores` (e.g., `host_config_daemon`)

File: src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    for field, value in fields.items():
+                        isEqual = isEqual and value == table[key][field]
+                        if not isEqual:
+                            break;
+                else:
+                    break
+        return isEqual
+
+    @parameterized.expand(hostcfgdTestVector)
+    def test_hostcfgd(self, name, testData):
```
Comment (jleveque) [inline src/sonic-host-services/tests/hostcfgd/hostcfgd_test.py:59]: Variable names in Python code should be `lowercase_with_underscores` (e.g., `test_data`)

File: src/sonic-host-services/tests/hostcfgd/test_vectors.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
+from unittest.mock import call
+
+"""
+    hostcfgd test vector
+"""
+hostcfgdTestVector = [
```
Comment (jleveque) [inline src/sonic-host-services/tests/hostcfgd/test_vectors.py:6]: Constant names in Python code should be `ALL_CAPS_WITH_UNDERSCORES` (e.g., `HOSTCFGD_TEST_VECTOR`)

File: src/sonic-host-services/.gitignore [hunk 5]
```diff
@@ -14,3 +14,4 @@ dist/
 .coverage
 coverage.xml
 htmlcov/
+/.pytest_cache/
```
Comment (jleveque) [inline src/sonic-host-services/.gitignore:17]: Leading slash should not be necessary; also keep list alphabetized:

```
.coverage
.pytest_cache/
coverage.xml
htmlcov/
```

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/sonic-host-services/.gitignore (hunk 5 above)
Comment (tahmed-dev) [reply src/sonic-host-services/.gitignore:17]: That was automatically added by the tool I am using.

Comment (jleveque) [reply src/sonic-host-services/.gitignore:17]: What tool is that?

Comment (tahmed-dev) [reply src/sonic-host-services/.gitignore]: I think it is `Git` plugin in `Eclipse`

File: src/sonic-host-services/.gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # Compiled Python files
+.pytest_cache/
```
Comment (jleveque) [inline src/sonic-host-services/.gitignore:2]: This belongs in the `# Unit test coverage` section if we're going to keep it. Please see my suggestion.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: src/sonic-host-services/.gitignore (hunk 6 above)
Comment (tahmed-dev) [reply src/sonic-host-services/.gitignore]: OK

**Review by lguohan — APPROVED** [review]

**Review by jleveque — APPROVED** [review]

**Pull request merged**